src/modm/communication/amnb/message.hpp

```diff
@@ -0,0 +1,157 @@
+enum class MessageType: uint8_t
+{
+	Broadcast = 0b00 << 6,	///< Broadcast message without response
+	Request   = 0b01 << 6,	///< Send request to a node
+	Error     = 0b10 << 6,	///< Respond to a request from a node with error code
+	Response  = 0b11 << 6,	///< Respond to a request from a node
+};
+
+/// @ingroup modm_communication_amnb
+struct Message //modm_packed
+{
+public:
+	inline uint8_t address() const          { return header_address; }
+	inline void setAddress(uint8_t address) { header_address = address; }
+
+	inline uint8_t command() const          { return header_command; }
+	inline void setCommand(uint8_t command) { header_command = command; }
+
+	inline uint16_t length() const
+	{
+		const uint8_t len = (header_type_length & LENGTH_MASK);
+		if (len == EXTENDED_LENGTH) return large.data_length;
+		return len;
+	}
+	inline void  setLength(uint16_t length)
+	{
+		if (length <= SMALL_LENGTH) {
+			header_type_length = (header_type_length & ~LENGTH_MASK) | length;
+		}
+		else {
+			header_type_length |= EXTENDED_LENGTH;
+			large.data_length = length;
+			allocateData();
+		}
+	}
+
+	inline MessageType type() const
+	{ return MessageType(header_type_length & ~LENGTH_MASK); }
+	inline void setType(MessageType type)
+	{ header_type_length = (header_type_length & LENGTH_MASK) | uint8_t(type); }
+
+	inline uint8_t* data() const
+	{ return isSmall() ? &small.data : large.data.get(); }
+
+public:
+	inline bool isHeaderValid() const
+	{ return crcHeader() == header_crc; }
+
+	inline bool isDataValid() const
+	{
+		if (not isHeaderValid()) return false;
+		if (dataLength()) return (crcData() == large.data_crc);
+		return true;
+	}
+
+	inline void setValid()
+	{
+		if (not isSmall()) large.data_crc = crcData();
+		header_crc = crcHeader();
+	}
+
+protected:
+	inline uint8_t
+	crcHeader() const
+	{
+		return modm::math::crc8_ccitt(&header_address, headerLength()-1);
+	}
+
+	inline uint16_t
+	crcData() const
+	{
+		return modm::math::crc16_ccitt(large.data.get(), dataLength());
+	}
+
+	inline bool isSmall() const
+	{
+		const uint8_t len = (header_type_length & LENGTH_MASK);
+		return len != EXTENDED_LENGTH;
+	}
+
+	inline uint8_t headerLength() const
+	{
+		const uint8_t len = (header_type_length & LENGTH_MASK);
+		return (len == EXTENDED_LENGTH) ? LARGE_HEADER_SIZE : (SMALL_HEADER_SIZE + len);
+	}
+
+	inline uint8_t dataLength() const
+	{
+		const uint8_t len = (header_type_length & LENGTH_MASK);
+		return (len == EXTENDED_LENGTH) ? large.data_length : 0;
+	}
+
+	inline bool allocateData(uint16_t max_size=0)
+	{
+		if (length() > SMALL_LENGTH) {
+			const size_t size = max_size ? std::min(length(), max_size) : length();
+			large.data.reset(new uint8_t[size]);
+		}
+	}
+
+protected:
+	// NOTE: This arrangement allows this data structure to be 4-byte aligned
+	//       so that any Cortex-M0 device has no issues with unaligned access
+	//       of neither the header values nor the data cast to another struct!
```

Comment:
Sadly, there is no way to make `char buffer[...] = {...}; reinterpret_cast<some_struct*>(buffer)->value` legal C++ yet. There is [p0593](http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2020/p0593r6.html) but the library part (`std::start_lifetime_as<T>()`) won't be in C++20 and will probably come in C++23. We can just hope no compiler optimizer will actually mess with the code until then...

Reply:
Hm, that's bad because that's exactly what I need to do here. I've added a basic implementation to create a union of `uint8_t[]` and a `std::shared_ptr<uint8_t[]>` that I'm reinterpret casting to variable types…

I'm starting to suspect that none of this is legal in C++, and it probably doesn't even do the correct thing with calling constructors/destructor etc. Is there any way to do this right? IIRC doing *anything* with the underlying memory of non-POD objects is dangerous, so "serializing" this data by simply copying the underlying memory is wrong?

Reply:
> I'm starting to suspect that none of this is legal in C++, and it probably doesn't even do the correct thing with calling constructors/destructor etc. Is there any way to do this right? IIRC doing anything with the underlying memory of non-POD objects is dangerous, so "serializing" this data by simply copying the underlying memory is wrong?

If your type is neither trivially default constructible or trivially copy constructible the whole approach is not sane. Either copying your object must be equivalent to copying bytes (memcpy) or any data must be a valid representation of your object.

A source of undefined behavior here is C++'s object model. You can't use an object that has not been created.

In C++ the following things create objects (from the standard draft)
> An object is created by
>   - a definition,
>   - by a new-expression,
>   -  when implicitly changing the active member of a union,
>   -  when a temporary object is created. 

and with the reinterpret_cast we do none of these.

For this reason, calling `malloc()`, casting the result to a POD struct and using it is strictly speaking undefined behaviour in C++ before 20. In C++20 this is solved with [p0593](http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2020/p0593r6.html) where they add that some operations implicitly create objects.

[placement new](http://www.cs.technion.ac.il/users/yechiel/c++-faq/placement-new.html) is probably the solution. But to do the placement new with the right type your code needs to know the type. This would imply some template things somewhere.

Reply:
Ok, thanks for the input. I've removed shared_ptr in favor of a simple manual memory manegement. I'm prematurely optimizing things here.

> copying your object must be equivalent to copying bytes (memcpy)

Yes, and it makes no sense to have it any other way for sending an obj state over wire. So fancy constructors are not possible anyways, even POD can be dangerous if they contain pointers. But this is something the user-facing API needs to check, perhaps even with Concepts (in the future).